### ISSUE TITLE
RNN | Recurrent Neural Network |循环神经网络

### ISSUE THREAD

**Issue by pillarliang** [pillarliang/blog#10]
# 背景

## Slot filling | 槽填充

`Slot filling` 是在给定的输入（用户的语音或文本输入）中识别并提取特定类型的实体或关键片段(`slot`)。换句话说，就是将这些关键信息填充到对应的槽位(`slot`)。

> Example: “*我想订一张下周一从北京飞往上海的机票*”
> 
> - “下周一” → “departure_date” 类型的槽位(slot)
> - “北京” → “departure_city”类型的槽位(slot)
> - “上海” → “arrival_city”类型的槽位(slot)
> 
> slot filling 将这些信息（例如，“下周一”、“北京”和“上海”）提取出来，并将它们填充到相关的slot（如“departure_date”、“departure_city”和“arrival_city”）
> 

## 前馈神经网络的问题

没有记忆能力，给定相同的输入，总会产生相同的输出。

> 例子：training：从北京出发到达上海。
> 
> 
> 输出一个概率分布，该概率分布代表着输入单词属于目的地槽的概率。
> 
> test：从上海出发到达北京。上海属于目的地槽的概率会大。
> 

# RNN

## What：

> RNN 就是有记忆单元的神经网络。循环指不断使用上一步的输出结果。在 **RNN** 里，隐藏层神经元的输出会被存到记忆元。考虑了远距离的words。
> 

### **基础架构**

$$
h_t = f(h_{t-1}, x_t)
$$

**输入输出：**

- 输入：基本的循环单元在每个时间步都有两个输入:$h^{t-1}, x^t$
- 输出：**y** 则常常使用 **h'** 投入到一个线性层（主要是进行维度映射）然后使用softmax进行分类得到需要的数据。

隐藏状态随时间传播信息

**参数共享**：在RNN的每个时间步中使用相同的权重和偏置参数。
![qw](https://github.com/pillarliang/blog/assets/28833267/6ad36d6a-db9a-4439-8012-384e71282466)

<img width="330" alt="qwe" src="https://github.com/pillarliang/blog/assets/28833267/28fc9ab3-494d-4843-87d1-f88f1347f068">

### Cost Function for RNNs

使用交叉熵损失函数。

RNN通常对一个序列进行预测，比如一句话有T个词。对每个时刻t，RNN都会输出一个预测,跟真实标注比较计算一个 loss。

最终 loss 就是简单的把T个时刻的 loss 加起来求平均。

### 双向循环神经网络

把正向的网络在输入 $x_t$ 的时候跟逆向的网络在输入 $x_t$ 时，都丢到输出层产生 $y_t$，产生 $y_{t+1}$, $y_{t+2}$，以此类推。

网络就等于整个输入的序列。假设考虑的是槽填充，网络就等于看了整个句子后，才决定每一个单词的槽，这样会比看句子的一半还要得到更好的性能。

<img width="500" alt="2w" src="https://github.com/pillarliang/blog/assets/28833267/8553615d-8d76-4308-864b-73d322fbaf00">

### 计算效率：

RNN是串联的，难以并行。

### 梯度消失

梯度是通过链式法则逐步[反向传播](https://www.notion.so/Backpropagation-2fc54e17c06d4bfa9c1e69914e84d09c?pvs=21)计算的，跨越多个时间步。在这个过程中，梯度会逐步衰减变小,这就是所谓的消失梯度问题。

梯度消失问题主要指的是对时间步靠前(离输出层较远)的梯度会越来越小，最后可能接近0。导致RNN对远距离依赖关系的学习不足,只能学习近距离特征。

<img width="549" alt="cd" src="https://github.com/pillarliang/blog/assets/28833267/32e67570-77a3-447f-986e-d711aa781b94">

**solution**

- 适当的激活函数：使用ReLU而不是sigmoid。
- 梯度裁剪：手动裁剪梯度值,控制其在一个定义好的阈值内,以防止梯度爆炸。
- 残差连接：添加跳跃连接或残差连接为梯度提供更短的反向传播路径。

### 梯度爆炸

同梯度消失，梯度消失是权重值在0-1之间。而梯度爆炸是权重大于1时。

# 改进

门控可以避免梯度消失和爆炸。允许网络有选择地决定什么信息应该被存储和遗忘。

1. **避免梯度消失**：
    - 由于门的设计，某些路径（例如LSTM的单元状态）允许梯度在多个时间步骤中几乎不受阻碍地传播。当网络决定保持某些信息时，这些信息可以在多个时间步内持续流动，从而避免了梯度消失问题。
2. **减轻梯度爆炸**：
    - 尽管门机制主要是为了解决梯度消失问题，但当梯度开始增长时，网络也可以利用门来选择性地遗忘或抑制某些信息，从而在某种程度上减轻梯度爆炸。

## Gated Recurrent Units(GRU) | 门控逻辑单元

![bg](https://github.com/pillarliang/blog/assets/28833267/c0841f45-ef1e-424d-8e58-f75141348235)

有两个门：

1. **更新门（Update Gate）**：决定了保留多少过去的信息。影响当前隐藏状态的最终输出。
2. **重置门（Reset Gate）**：应该忽略多少之前的隐藏状态的信息。在计算候选隐藏状态时发挥作用。

> $h_t = f(h_{t-1}, x_t)$:
> 
> 
> 重置门就是决定 $h_{t-1}$ 用不用，用多少。
> 
> 更新门就是决定 $h_t$ 输出多少。
> 

## LSTM | Long short-term memory unit

> 为了解决长序列训练过程中的梯度消失和梯度爆炸问题。
> 

<img width="433" alt="3c" src="https://github.com/pillarliang/blog/assets/28833267/eae47cfe-b8b5-4f16-8b49-3893c75e9b02">

### 结构：

LSTM有**两个传输状态**：

- $c^t$（cell state）表示LSTM单元内部的记忆信息,可以跨时间步传递。
- $h^t$（hidden state）表示LSTM在当前时间步的输出特征。

**三个门**（输入、遗忘和输出门）和**一个记忆单元**：

- 输入门：决定要存储的信息
- 输出门：决定当前步骤要使用的信息
- 遗忘门：决定哪些信息不再重要
    
    遗忘门的开关是跟直觉相反，*打开* 代表记得，*关闭* 代表遗忘。
    
![1112](https://github.com/pillarliang/blog/assets/28833267/f1cf7a5a-512c-4a9c-ad92-76b38ef74e36)

<img width="480" alt="12341" src="https://github.com/pillarliang/blog/assets/28833267/8c1be60c-32fc-47ba-bdb2-04fdc51bbdfc">

<img width="499" alt="444" src="https://github.com/pillarliang/blog/assets/28833267/b0c90139-97fb-4b6b-803a-836bcf9adb2d">

操控 LSTM 四个门的时候，同时考虑了 $x_{t+1}$,  $h_t$, $c_t$，把这三个向量并在一起乘上不同的变换得到四个不同的向量再去操控 LSTM。

### 使用：

把 LSTM 想成是一个神经元。所以要用一个 LSTM 的神经元，其实就是原来简单的神经元换成 LSTM。

<img width="361" alt="fr" src="https://github.com/pillarliang/blog/assets/28833267/4b3457c5-e80b-44dc-909c-c34b0662c43d">

# Code

- nn.RNN 类初始化
    - `input_size`: 输入张量 x 的维度
    - `hidden_size`: 隐层张量 h 的维度
    - num_layers: hidden layer 的数量
    - nonlinearity: 激活函数的选择, 默认是tanh
    
    ```python
    self.rnn = nn.RNN(input_size, hidden_size, num_layers)
    ```
    
- nn.RNN类实例化对象主要参数解释:
    - 输入参数：
        - input: 输入张量
        - h_0: 记忆单元的输入，初始为0
    - 输出参数：
        - output: 网络层输出
        - h_n: 记忆单元输出/隐藏层输出
    
    ```python
    output, hn = self.rnn(input, h_0)
    ```
